### DISCUSSION TITLE
Avalonia and Raspberry

### DISCUSSION THREAD

**Discussion by Robiberto** [AvaloniaUI/Avalonia#17039] [Q&A] [answered]
Dear Forum.
My company, need to develop GUI on window, and, on Raspberry. Im considering to use Avalonia.
Question are:
1)Is possible to use Avalonia also for raspberry ( i means cross compilation ..install VS on window or linux, and deploy or compile on Rasp) ?
2) I read that is possible to avoid XAML for GUI design; is this also true for Raspberry APP  (if the reply 1 is positive).

Thanks
Roberto

**Comment by thevortexcloud** [comment]
> )Is possible to use Avalonia also for raspberry ( i means cross compilation ..install VS on window or linux, and deploy or compile on Rasp) ?

Avalonia supports desktop Linux as well as Framebuffer/DRM mode. Any of those will work on a Raspberry Pi. You cannot install Visual Studio on Linux though. The only real modern C# IDE that exists for Linux is Rider, which _does_ work on a Raspberry Pi. But don't expect the performance to be great. If you just want to compile though on the Pi, you can do that from a terminal with only a couple commands.

If you want to cross compile from Windows to Linux, the answer is: It depends. For a normal .NET app, it will work fine. An AOT app however requires lots of fiddling with obscure packages and build settings which is generally not worth it if you can just build on the actual hardware. You can use WSL for some of it, but you will still run into problems with the Pi using a different architecture from a regular PC.

Visual Studio (and Rider) also has the ability to publish over SSH, so you can easily remote deploy from a Windows PC to the Pi.

>  I read that is possible to avoid XAML for GUI design; is this also true for Raspberry APP (if the reply 1 is positive).

It's possible for all Avalonia apps. I would not recommend it though. It's not as well documented and some things are significantly harder to do. But there is this library to make it easier:

https://github.com/AvaloniaUI/Avalonia.Markup.Declarative

**Reply by Robiberto** [reply]
Thanks for your reply.

Excuse if I maybe reply with same question. 
Time ago I develop a linux app with Visual Studio,connect via SSH to the targhet, buld, and all was ok. This targhet was a  device with Linaro, but also a Raspberry.
All was ok.
But here i would like to use Avalonia and the .NET8.  I read the .NET 8 is cross platform, and include Linux, and, with Avalonia i have access to some WPF ,and so, 
i will be able to design some GUI . So:

1) when i didi this on window PC, can I connect via SSH to rasp,and compile on it
2)of course, I need to install .NET 8 on rasp 

Excuse again for repeat same question.

Thanks
Roberto

**Reply by thevortexcloud** [reply]
> with Avalonia i have access to some WPF ,and so,

Avalonia is _not_ WPF. It's a completely different, albeit very similar thing. If you want to use WPF on Linux you need to use XPF, which is a paid product by Avalonia.

https://avaloniaui.net/xpf

.NET 8 itself works fine on Linux. Some of the APIs don't work outside Windows/don't behave the same for no obvious reason but the vast majority of it is fine.

.NET code itself is actually platform agnostic. It all gets compiled CIL and then the native .NET runtime works out what to do with it.

https://en.wikipedia.org/wiki/Common_Intermediate_Language

You can compile a .NET/Avalonia program on a Windows machine and execute it on a Linux one without issue. You can also do the reverse if you want. You run into problems with using platform specific APIs (such as WPF which only works on Windows) and native libraries (such as those Avalonia uses) which need to be built for the system they are running on. But you can work around most of that by setting up your publish settings properly.

I personally have made many apps on Linux using Avalonia, and also a few targeting various different Raspberry Pis. There are _some_ Raspberry Pis (mainly very old ones) you can't use moden .NET on due to the runtime not supporting the hardware. You can however use Mono, which I think Avalonia supports?

**Reply by Robiberto** [reply]
Thanks again for reply .
 
Clear about WPF : it is window specific,so, no portable on other system. OK.
Clear also about ..native lib..  You means, for example, if I include WinForm,  it is not portable.

If possible, I ask 2 question again, before conclusion:

1) The app we need to develop on Rasp and window, need some buttons, file access, and plot XY
 graph; for this, could be possible using only Avalonia framework and .NET8....No need for some native lib.  Correct ?

2)On the targhet machine (linux or rasp),i need to install the .NET8 and Avalonia .So, via SSH we ca transfer the sorce code, and compile on targhet.  Is this correct ?


Thanks again very much

Roberto

**Reply by thevortexcloud** [reply, marked as answer]
> The app we need to develop on Rasp and window, need some buttons, file access, and plot XY
graph; for this, could be possible using only Avalonia framework and .NET8....No need for some native lib. Correct ?

You still need a native library. Avalonia relies heavily on things like Skia for basically everything. The .NET build system (if it's configured to do so) generally works out what libraries to include based on what platform you target. 

> i need to install the .NET8 and Avalonia .So, via SSH we ca transfer the sorce code, and compile on targhet. Is this correct ?

Yes but also no. Again, it depends on how you compile the app. If you do a self contained app then the app will contain the enitre .NET runtime. Similarly, if you AOT compile it will instead be converted into native code. If you do a framework dependent compile however, you will need to install the runtime on the actual device.

If you specifically want to build on the target though, then you would need to install the .NET SDK.

**Reply by Robiberto** [reply]
Thanks very very much for reply.

Roberto
